packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -2885,11 +3091,131 @@ function remountFiber(
   }
 }
 
+function reifyWorkInProgress(current: Fiber, workInProgress: Fiber | null) {
+  didReifySpeculativeWork = true;
+
+  if (workInProgress === null) {
+    // intentionally not returning anywhere yet
+    workInProgress = createWorkInProgress(
+      current,
+      current.pendingProps,
+      NoWork,
+    );
+    workInProgress.return = null;
+  }
+  let originalWorkInProgress = workInProgress;
+
+  invariant(
+    speculativeWorkRootFiber !== null,
+    'reifyWorkInProgress was called when we are not in speculative work mode',
+  );
+
+  while (true) {
+    // fibers in the part of the tree already visited by beginWork need to have
+    // their expirationTime set to NoWork. fibers we have yet to visit in the reified
+    // tree need to maintain their expirationTime. This might need to be reworked
+    // where we don't reify to-be-visited fibers but that requires tracking speculative
+    // mode on the fiber.mode property
+    let alreadyCompletedSpeculativeChildFibers = true;
+
+    // when we encounter the reifiedCurrentChild we need to assign the reifiedNewChild rather
+    // than construct a new workInProgress
+    let reifiedNewChild = workInProgress;
+    let reifiedCurrentChild = current;
+
+    // special logic required when we are reifying into the speculativeWorkRootFiber
+    // this is because current.return will actually be pointing at the last workInProgress
+    // created before speculative work began whereas in other cases it will be pointing
+    // to it's parent from the current tree
+    if (current.return === speculativeWorkRootFiber) {
+      workInProgress = current.return;
+      current = current.return = workInProgress.alternate;
+    } else {
+      current = current.return;
+      workInProgress = createWorkInProgress(
+        current,
+        current.pendingProps,
+        NoWork,
+      );
+      workInProgress.return = null;
+    }
+
+    // visit each child of workInProgress and create a newChild
+    let child = workInProgress.child;
+    let parent = workInProgress;
+    let newChild = voidFiber;
+    while (child !== null) {
+      if (child === reifiedCurrentChild) {
+        newChild = newChild.sibling = parent.child = reifiedNewChild;
```

Comment:
using a technique like voidFiber allows this heavily linearized assignment. it's probably too cute and I have no idea if it actually matters for perf. if we didn't use it we'd have to have an additional conditional to treat wip.child different than child.sibling